[WIP] 491 formal spec reorg (#679)
finos/FDC3: docs/app-directory/overview.md

The application metadata stored in appD records may include: the app name, type, details about how to run the application, its icons, publisher, support contact details and so on. It may also include links to or embed manifest formats defined elsewhere, such as proprietary manifests for launching the app in a container product or a Web Application Manifest (as [defined by the W3C](https://www.w3.org/TR/appmanifest/)).

All this information is readily available in one place and can be used both to populate a launcher or app catalog UI for your users, and by the desktop agent managing the apps on your desktop. In fact, if your desktop platform supports the FDC3 standard, appD is the primary way that the FDC3 desktop agent implementation should receive the details about apps available to run on your desktop. Conversely, if an app is not listed in appD, the desktop agent can’t ensure its participation in context sharing or use it to resolve intents.

## Advantages

> Using appD offers many advantages both for the financial institutions running an FDC3-enabled desktop container and for vendors that provide FDC3-compliant apps:

Reviewer: ```suggestion
Using appD offers many advantages both for financial institutions running an FDC3-enabled desktop container and for vendors that provide FDC3-compliant apps:
```

#### Easy Updates

As a vendor, you prefer for all your customers to run your latest software. However, many customers will postpone upgrades, sometimes for a long time, because upgrading can be a pain. An advantage of a vendor-hosted appD is that the configuration of an app can be updated at any time and, if your customers need to choose when to upgrade, multiple versions of it can be made available, each with their own configuration. By making it easier for customers to update, you can drive better adoption.

### For the Industry

By hosting our own appD we can easily combine applications from various providers into one cohesive directory. Alternatively, we can connect to directories from multiple providers (in standardized format) and provide a single view over them. This reduces fragmentation in the market, allows end-users more flexibility in what apps to include in their smart desktop, and obviates the need for vendors to provide application details in diverse formats or for their customers to work out these details for themselves.

## Use Cases

> An application directory (AppD) provides information about an application's identifiers, publisher details, intents that it supports, and metadata necessary to launch and integrate the application in a desktop agent.

Reviewer: Until this point on the page the 'a' in 'appD' has been lower case. I don't really have a preference but we should be consistent.

Reviewer: it seems to be all 'AppD' after this

Author: Duly noted - product of merging different docs. We should always use whatever capitalisation is defined in the glossary (now that we have one), where it is lowercase.
```suggestion
An application directory provides information about an application's identifiers, publisher details, intents that it supports, and metadata necessary to launch and integrate the application in a desktop agent.
```

The following provides a summary of use cases.

### Launcher

A desktop agent will usually include a user interface allowing the user to select from a set of launchable applications and then allow them to manually launch an one. It is also responsible for launching applications necessary to resolve a raised intent. However, it must first retrieve the necessary metadata about the available applications. An app directory provides an endpoint to retrieve a list of the available applications along with their metadata, which may include or link to additional information necessary to launch the application in a specific desktop agent.

> A launcher will usually be configured with the locations of one or more AppD servers (which is necessary to implement intent resolution), however, as described in the [Service Discovery](#service-discovery) section, a fully qualified application identifier (app1@host.appd.com) may also be used to both locate the AppD service and to retrieve the specific application data.

Author: ```suggestion
A launcher will usually be configured with the locations of one or more AppD servers (which is necessary to implement intent resolution), however, as described in the [Service Discovery](#service-discovery) section, a fully qualified application identifier (app1@host.appd.com) may also be used to both locate the appD service and to retrieve the specific application data.
```

![img](assets/appd_launcher_embedded.png)

### Aggregated View

> There could be many different AppD service instances in the world providing application data zoned to the provider or enterprise deployment. The AppD specification allows for unique instances of the service with no requirement to aggregate data or define a structured hierarchy. With this said, a launcher might want to construct an aggregated view of applications from one or more AppD instances. In this case, the launcher would be required to retrieve multiple application definitions from one or more AppD instances providing a consolidated view of all applications required.

Author: ```suggestion
There could be many different appD service instances in the world providing application data zoned to the provider or enterprise deployment. The appD specification allows for unique instances of the service with no requirement to aggregate data or define a structured hierarchy. With this said, a launcher might want to construct an aggregated view of applications from one or more appD instances. In this case, the launcher would be required to retrieve multiple application definitions from one or more appD instances providing a consolidated view of all applications required.
```

Today, there is no intention to create a single registry of known AppD instances, so there is an assumption that the launcher will have prior
knowledge of the AppD instance location.

![img](assets/appd_launcher_aggregated.png)

### Authentication and Entitlements

> The AppD API specification defines the optional use of an access token to identify the requesting user/launcher and implement authorizations which may affect AppD API responses. For example, different subsets of the full list of applications may be returned for different users depending on their role in an organization.  

Author: ```suggestion
The AppD API specification defines the optional use of an access token to identify the requesting user/launcher and implement authorizations which may affect appD API responses. For example, different subsets of the full list of applications may be returned for different users depending on their role in an organization.  
```

## Service Discovery

In order to support the discovery of applications that can be used with a desktop agent, it is necessary to access data stored in one or more app directory instances. 

![img](assets/appd_service_distribution.png)

However, in order to do so, you must first discover the location of an app directory service, which you may then use to generate URIs  (e.g. "https://appd.foo.com/api/appd/v2/app1@appd.foo.com") to query a given directory instance for data. In order to construct a URI, the host location and port of a given AppD service instance is required.

Three methods for discovering app directory services are defined in this Standard:

  1. **Static configuration:** Statically defined URI records for use within client applications (typically a desktop agent implementation) directly. 

>   2. **Fully-qualified appID namespace syntax host resolution:** Discovery of the AppD location using a fully qualified application ID (appId) domain name.  

Author: ```suggestion
  2. **Fully-qualified appID namespace syntax host resolution:** Discovery of the appD location using a fully qualified application ID (appId) domain name.  
```

  3. **DNS lookup by domain name:** Discovery of the AppD location using a domain name to lookup DNS SRV records identifying the host server location and TCP port. ([RFC2782](https://tools.ietf.org/html/rfc2782))

> AppD service host discovery implementations SHOULD support each of these methods and MUST support at least static configuration.

Author: ```suggestion
App directory service host discovery implementations SHOULD support each of these methods and MUST support at least static configuration.
```

### Static configuration

> As the name implies, a static configuration for an AppD service location is defined within a desktop agent, launcher application. This is the simplest and most common approach to app directory and application data discovery.

Author: ```suggestion
As the name implies, a static configuration for an appD service location is defined within a desktop agent, launcher application. This is the simplest and most common approach to app directory and application data discovery.
```

![img](assets/appd_static_config.png)

### Fully-qualified appID namespace syntax host resolution

> An application directory URI can be constructed using a [fully qualified application ID](spec#applicationidentifiers) (email address syntax) by using fqdn part of the ID as the host location and the name part as the application name. Given an application id "app1" with a fully qualified identifier of "app1@appd.foo.com" an application directory host location can be derived by simply extracting the fqdn "appd.foo.com" from the email syntax. The extracted fqdn "app.foo.com" MUST resolve to the actual host location where the application directory is running.

Author: ```suggestion
An app directory URI can be constructed using a [fully qualified application ID](spec#applicationidentifiers) (email address syntax) by using fqdn part of the ID as the host location and the name part as the application name. Given an application id "app1" with a fully qualified identifier of "app1@appd.foo.com" an application directory host location can be derived by simply extracting the fqdn "appd.foo.com" from the email syntax. The extracted fqdn "app.foo.com" MUST resolve to the actual host location where the application directory is running.
```

A launcher can then easily construct a URI by:

1. URI protocol is defaulted to `https`, but can be overridden by the launcher.
2. URI hostname is the fully qualified domain of the application ID. 
3. URI port is default `https/443`, but can be overridden by the launcher
4. URI url is by default `"/api/appd/(version)"` . Calls that are made without version MUST automatically default to latest, i.e. `"/api/appd/app1"` should return the same result as `"/api/appd/v2/app1"`.

The resulting URI to retrieve application data for "app1" would be "[https://appd.foo.com/api/appd/v2/app1@appd.foo.com](https://appd.foo.com/api/appd/v2/app1@appd.foo.com)"  

### DNS/SRV Records

> Another approach to support AppD service discovery (resolution) is through use of existing domain name service (DNS) implementations that are broadly used on the Internet today (see: [RFCs](https://www.isc.org/community/rfcs/dns/)). Name service implementations can be considered critical infrastructure and are proven stable with over twenty years of use. Name services can be used both through public Internet or locally deployed intranet, which provides optionality to deployment schemes.

Author: ```suggestion
Another approach to support app directory service discovery (resolution) is through use of existing domain name service (DNS) implementations that are broadly used on the Internet today (see: [RFCs](https://www.isc.org/community/rfcs/dns/)). Name service implementations can be considered critical infrastructure and are proven stable with over twenty years of use. Name services can be used both through public Internet or locally deployed intranet, which provides optionality to deployment schemes.
```

> More specifically, resolution of an AppD service instance (host location) can be implemented using DNS "service records" (SRV) providing the host instance, protocol and associated port. The following is a well known description of a SRV record ([RFC2782](https://tools.ietf.org/html/rfc2782)):

Author: ```suggestion
More specifically, resolution of an appD service instance (host location) can be implemented using DNS "service records" (SRV) providing the host instance, protocol and associated port. The following is a well known description of a SRV record ([RFC2782](https://tools.ietf.org/html/rfc2782)):
```

- *name*: the domain name for which this record is valid, ending in a dot.  For AppD service,  the name should directly map to the application identifier domain.
- *TTL*: standard DNS [time to live](https://en.wikipedia.org/wiki/Time_to_live) field.
- *class*: standard DNS class field (this is always *IN*).
- *priority*: the priority of the target host, lower value means more preferred.
- *weight*: A relative weight for records with the same priority, higher value means more preferred.
- *port*: the TCP or UDP port on which the service is to be found. For AppD service, TCP should always be used. 
- *target*: the canonical hostname of the machine providing the service, ending in a dot. This would be the host where the AppD service is running. 

For AppD Service the SRV record MUST use the following definitions:

- service = **_appd**
- proto = **_tcp**
- name = must map to the domain of the application identifier . Example:  the **name** for application identifier **"app1@appd.foo.com"** would be **"appd.foo.com"**

*AppD service through DNS / SRV records:*

![img](assets/appd_dns.png)

#### Known domains

> Although SRV records provide the means of resolving the location of an AppD service for a specific domain, there could be a need to know what domains exist in the universe. This would be a list of domains representing all known directory instances. It is recommended that  the FDC3/FINOS organization publish a list of known domains which support AppD services. This publication can be handled in multiple ways, such as structured files or API endpoints. This proposal shall not provide a qualified solution to achieve this, but rather draw attention to a potential requirement.

Author: ```suggestion
Although SRV records provide the means of resolving the location of an app directory service for a specific domain, there could be a need to know what domains exist in the universe. This would be a list of domains representing all known directory instances. It is recommended that the FDC3/FINOS organization publish a list of known domains which support AppD services. This publication can be handled in multiple ways, such as structured files or API endpoints. This proposal shall not provide a qualified solution to achieve this, but rather draw attention to a potential requirement.
```
